js/compiler.ts

```diff
@@ -0,0 +1,586 @@
+      containingFile.startsWith(ASSETS)
+    ) {
+      // Assets are compiled into the runtime javascript bundle.
+      // we _know_ `.pop()` will return a string, but TypeScript doesn't so
+      // not null assertion
+      const moduleId = moduleSpecifier.split("/").pop()!;
+      const assetName = moduleId.includes(".") ? moduleId : `${moduleId}.d.ts`;
+      assert(assetName in assetSourceCode, `No such asset "${assetName}"`);
+      sourceCode = assetSourceCode[assetName];
+      fileName = `${ASSETS}/${assetName}`;
+    } else {
+      // We query Rust with a CodeFetch message. It will load the sourceCode,
+      // and if there is any outputCode cached, will return that as well.
+      let fetchResponse;
+      try {
+        fetchResponse = this._os.codeFetch(moduleSpecifier, containingFile);
+      } catch (e) {
+        return throwResolutionError(
+          `os.codeFetch message: ${e.message}`,
+          moduleSpecifier,
+          containingFile
+        );
+      }
+      fileName = fetchResponse.filename || undefined;
+      sourceCode = fetchResponse.sourceCode || undefined;
+      outputCode = fetchResponse.outputCode || undefined;
+    }
+    if (!sourceCode || sourceCode.length === 0 || !fileName) {
+      return throwResolutionError(
+        "Invalid source code or file name.",
+        moduleSpecifier,
+        containingFile
+      );
+    }
+    this._log("resolveModule sourceCode length ", sourceCode.length);
+    this.setFileName(moduleSpecifier, containingFile, fileName);
+    if (fileName && this._moduleMetaDataMap.has(fileName)) {
+      return this._moduleMetaDataMap.get(fileName)!;
+    }
+    const moduleMetaData = new ModuleMetaData(fileName, sourceCode, outputCode);
+    this._moduleMetaDataMap.set(fileName, moduleMetaData);
+    return moduleMetaData;
+  }
+
+  /**
+   * Resolve the `fileName` for a given `moduleSpecifier` and `containingFile`
+   */
+  resolveModuleName(
+    moduleSpecifier: ModuleSpecifier,
+    containingFile: ContainingFile
+  ): ModuleFileName | undefined {
+    const moduleMetaData = this.resolveModule(moduleSpecifier, containingFile);
+    return moduleMetaData ? moduleMetaData.fileName : undefined;
+  }
+
+  /* tslint:disable-next-line:no-any */
+  /**
+   * Execute a module based on the `moduleSpecifier` and the `containingFile`
+   * and return the resulting `FileModule`.
+   */
+  run(
+    moduleSpecifier: ModuleSpecifier,
+    containingFile: ContainingFile
+  ): ModuleMetaData {
+    this._log("run", { moduleSpecifier, containingFile });
+    const moduleMetaData = this.resolveModule(moduleSpecifier, containingFile);
+    const fileName = moduleMetaData.fileName;
+    this._scriptFileNames = [fileName];
+    const sourceCode = moduleMetaData.sourceCode;
+    let outputCode = moduleMetaData.outputCode;
+    if (!outputCode) {
+      outputCode = moduleMetaData.outputCode = `${this.compile(
+        fileName
+      )}\n//# sourceURL=${fileName}`;
```

Comment:
(This is a comment that you don't need to address before landing.)

When a module is compiled for the first time, we create moduleMetaData instance with undefined outputCode. The moduleMetaData is basically our local VM cache of the outputCode. Now, here we update the outputCode - but this only for main scripts. Modules which are loaded as imports don't seem to get outputCode update. Could this be the cause of the superfluous codeFetch calls?